### ISSUE TITLE
forward plugin issues with truncated responses without force_tcp or prefer_udp config

### ISSUE THREAD

**Issue by kishorj** [coredns/coredns#4183]
**What happened**:
When either prefer_udp or force_tcp options are not set in the forward config, CoreDNS follows the client protocol when talking to the upstream servers. If a client sent DNS request over UDP without the EDNS0 option, and the upstream server sent a truncated response, CoreDNS doesn't retry the request over TCP. It also doesn't set the TC flag in the response sent to the client. As a result, client ends up with a poetntially incomplete DNS response, and potentially fail to resolve the hostname permanently.

**What you expected to happen**:
In case the client sent requests over UDP and force_tcp or perfer_udp option is not set in the config, handle the truncated responses by either retrying over TCP or pass on the TC flag to the client.  

**How to reproduce it (as minimally and precisely as possible)**:
Use forward config without force_tcp or prefer_udp option, dns client use UDP, and not set the EDNS0 option. For example
```
dig +noedns site.with.truncated.response
```
**Anything else we need to know?**:

**Environment**:

- the version of CoreDNS: 1.6.4, 1.6.6, 1.6.7, 1.7.0
- Corefile:
```
  Corefile: |
    .:53 {
        errors
        health
        kubernetes cluster.local in-addr.arpa ip6.arpa {
          pods insecure
          fallthrough in-addr.arpa ip6.arpa
        }
        prometheus :9153
        forward . /etc/resolv.conf
        cache 30
        loop
        reload
        loadbalance
    }
```
- logs, if applicable:
- OS (e.g: `cat /etc/os-release`):
- Others:

**Comment by chrisohaver** [comment]
> `dig +noedns site.with.truncated.response`

dig will automatically retry with tcp if the response is truncated. When it does, it doesn't show the intermediate truncated response.  If you use the `+ignore` flag, it will not retry, and display the truncated response.

**Comment by chrisohaver** [comment]
I verified that CoreDNS passes along truncated responses, with the TC bit set to the client.  I tested with CoreDNS on current master branch configured to run 2 servers, one forwarding to the other using template to create a large response.

**Comment by kishorj** [comment]
> > `dig +noedns site.with.truncated.response`
> 
> dig will automatically retry with tcp if the response is truncated. When it does, it doesn't show the intermediate truncated response. If you use the `+ignore` flag, it will not retry, and display the truncated response.

I see that dig automatically retry with TCP for truncated response when CoreDN is not in the path. I also verified with the packet captures. In case of CoreDNS, if forward config doesn't contain force_tcp or prefer_udp, I see CoreDNS doesn't seem handle the truncated flag correctly. I verified this with versions 1.6.4, 1.6.6, 1.6.7, 1.7.0.

**Comment by kishorj** [comment]
> I verified that CoreDNS passes along truncated responses, with the TC bit set to the client. I tested with CoreDNS on current master branch configured to run 2 servers, one forwarding to the other using template to create a large response.

Things get a bit more interesting with the latest builds from the master branch. I see that CoreDNS adds the EDNS0 OPT record unconditionally. Due to the larger buffer, the response from upstream is no longer truncated, and my previous tests work fine. However, I see that CoreDNS adds the OPT record to the client response even when client doesn't indicate edns0 support in the query

```
$ dig +noedns site.with.truncated.response @localhost

; <<>> DiG 9.10.6 <<>> +noedns site.with.truncated.response @localhost
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 466
;; flags: qr rd ra; QUERY: 1, ANSWER: 14, AUTHORITY: 4, ADDITIONAL: 5

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags: do; udp: 2048
;; QUESTION SECTION:
```

Here is the configuration - 
```
.:53 {
        errors
        forward . /etc/resolv.conf
        cache 30
}

```
Cached lookups seem fine, only non-cached lookup exhibit this behavior.

This new behavior is a different issue and sending larger responses to non-edns0 clients could potentially cause them to malfunction. This is also a protocol violation.

As for the truncated flag handling, please confirm that the CoreDNS from the current master branch, without force_tcp or prefer_udp in the forward config, is able to handle upstream responses that are well formed but contain the TC flag set.

**Comment by chrisohaver** [comment]
>  I verified this with versions 1.6.4, 1.6.6, 1.6.7, 1.7.0.

OK, I verified above that it works correctly in the master.  I had tested with and without cache as well.  The cache plugin adds some variability in the response (sometimes response contained truncated set of answers, other times it contained no answers), but CoreDNS still always replied with the TC bit set. So it seems that either something was fixed after 1.7.0, or there is some other difference in our test procedure/expectation.


Can you share step by step example/demonstration of dig results showing the truncated response without TC bit set?
Edit - i see you did, as i was writing this... thanks.

**Comment by chrisohaver** [comment]
OK - i verified my test again, and it works as expected regarding the TC bit.  With cache - I do see edns enabled in the response - I agree i think this is a separate issue.

However, I varied my test a little, and now I _can_ reproduce a truncated response _without_ a TC bit.
Below is a step by step repro ... (note: the response to the first query `123456.` has the TC bit set, the response to `1234567.` does not) ...

```
$ cat Corefile.template 
.:5399 {
    forward . 127.0.0.1:5398
}

.:5398 {
    errors
    debug
    template ANY ANY . {
      answer "{{ .Name }} 60 IN TXT 0-0123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789"
      answer "{{ .Name }} 60 IN TXT 1-0123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789"
      answer "{{ .Name }} 60 IN TXT 2-0123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789"
      answer "{{ .Name }} 60 IN TXT 3-0123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789"
      answer "{{ .Name }} 60 IN TXT 4-0123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789"
      answer "{{ .Name }} 60 IN TXT 5-0123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789"
      answer "{{ .Name }} 60 IN TXT 6-0123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789"
      answer "{{ .Name }} 60 IN TXT 7-0123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789"
      answer "{{ .Name }} 60 IN TXT 8-0123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789"
      answer "{{ .Name }} 60 IN TXT 9-0123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789"
    }
}

$ ./coredns -conf Corefile.template 
.:5399
.:5398
CoreDNS-1.7.1
darwin/amd64, go1.15.2, f2ec292e-dirty


$ dig +noedns @localhost -p5399 -t TXT -b 127.0.0.1 123456. +ignore

; <<>> DiG 9.16.5 <<>> +noedns @localhost -p5399 -t TXT -b 127.0.0.1 123456. +ignore
; (2 servers found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 25232
;; flags: qr aa tc rd; QUERY: 1, ANSWER: 4, AUTHORITY: 0, ADDITIONAL: 0
;; WARNING: recursion requested but not available

;; QUESTION SECTION:
;123456.				IN	TXT

;; ANSWER SECTION:
123456.			60	IN	TXT	"0-0123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789"
123456.			60	IN	TXT	"1-0123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789"
123456.			60	IN	TXT	"2-0123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789"
123456.			60	IN	TXT	"3-0123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789"

;; Query time: 0 msec
;; SERVER: 127.0.0.1#5399(127.0.0.1)
;; WHEN: Thu Oct 08 13:21:25 EDT 2020
;; MSG SIZE  rcvd: 508

$ dig +noedns @localhost -p5399 -t TXT -b 127.0.0.1 1234567. +ignore

; <<>> DiG 9.16.5 <<>> +noedns @localhost -p5399 -t TXT -b 127.0.0.1 1234567. +ignore
; (2 servers found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 31987
;; flags: qr aa rd; QUERY: 1, ANSWER: 4, AUTHORITY: 0, ADDITIONAL: 0
;; WARNING: recursion requested but not available

;; QUESTION SECTION:
;1234567.			IN	TXT

;; ANSWER SECTION:
1234567.		60	IN	TXT	"0-0123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789"
1234567.		60	IN	TXT	"1-0123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789"
1234567.		60	IN	TXT	"2-0123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789"
1234567.		60	IN	TXT	"3-0123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789"

;; Query time: 0 msec
;; SERVER: 127.0.0.1#5399(127.0.0.1)
;; WHEN: Thu Oct 08 13:21:30 EDT 2020
;; MSG SIZE  rcvd: 485
```

The total size of the un-truncated message seems to be a factor. I'll run this though the debugger and try to see what's going on.
